lib/logstash/filters/jdbc/lookup.rb

```diff
@@ -60,44 +69,69 @@ def formatted_errors
+    # @param [ReadWriteDatabase] local, the local derby database
+    # @param [LogStash::Event] event, the current event being enhanced by this lookup
+    # @return [Boolean] returns true when the event contains the looked up fields/values or the default value was applied. False in all other cases
+    # meaning that only if all lookups enhance successfully then `filter_matched` is called
     def enhance(local, event)
+      # LookupFailures API (JRuby extension)
+      # successful?,  no errors, event was updated with looked up fields/values
+      # invalid_parameters, array of invalid parameters
+      # checking_columns?, has valid/invalid columns been collected for this instance
+      # check_columns, we only need to do this once
+      # all_columns, array of columns attempted to convert to Ruby, if collected, this array is converted from a java HashSet to ruby on demand
+      # invalid_columns, array of unacceptable column datatype messages, if collected, this array is converted from a java HashSet to ruby on demand
+      # invalid_id_for_lookup?, was the id supplied unregistered? Very unlikely.
+      # any_invalid_columns?, are there any unacceptable columns in the resultset?
+      # any_invalid_parameters?, are there any parameters that did not...
+      #      A) interpolate well
+      #      B) resolve to nil, array or hash,
+      #      C) have a type we don't coerce to a SQL type (unlikely, as we cover what Valuefier handles)
       begin
-        result = local.fetch_with_lock(@id, event) # returns an instance of LookupFailures, defined in JRuby extension
-      rescue StandardError
+        result = if @column_validity_checked.true?
+                  LookupFailures.new
+                else
+                  @column_validity_checked.make_true
+                  LookupFailures.new.check_columns
+                end
+        local.fetch_with_lock(@id, event, result) # updates the instance of LookupFailures (see comments above), defined in JRuby extension
+        return true if result.successful?
+      rescue StandardError => e
         # a SQLException re-thrown as as a StandardError, the cause and backtrace has been logged already in Java
         # this error holds no meaningful info.
         tag_failure(event)
         return tag_and_set_default(event)
       end
 
+      if result.any_invalid_parameters?
+        logger.warn("Parameters for the statement cannot be prepared, interpolation may have failed or the value might be nil, array or hash", "parameters" => result.invalid_parameters, "event" => event.to_hash)
+        # if this is a checking columns run and there are invalid parameters then the prepared statement is not
+        # even executed so columns cannot validated this time - need to check columns again on next event.
+        @column_validity_checked.make_false if result.checking_columns?
```

Comment:
Is there a case where setting this value back to false may cause an event to 'fall through the cracks' and miss the column check? I wonder if we don't set this to true unless the columns have definitely been checked, either through a successful return or `any_invalid_columns?` being `true`. Downside - potentially some redundant logging, anything else?

Reply:
There is.

Another thread with its first event will take the `if  @column_validity_checked.true?` path and call into the Java code with column checking off. The call may complete earlier due to this - that event will not have columns checked.

Remedies:
A) Always check columns regardless.
B) Block other threads until a column check is successful.

I'm leaning towards `A`. The overhead is not so great that it warrants the deadlock risk and complexity that `B` implies.

Reply:
I would tend to agree that you don't want to block until the column check is successful. The other option is to do the check until the column check is completed, instead of always, although I don't whether the trade off between code complexity and performance is worth the effort